File: 14-longest-common-prefix/14-longest-common-prefix.py
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        def findPre(left, right):
            pre = ""
            for i in range(min(len(left), len(right))):
                if left[i] == right[i]:
                    pre += left[i]
                else:
                    break
            return pre
        
        def divideAndConcur(start, end):
            if start == end:
                return strs[start]
            mid = start + (end - start) // 2
            left_half = divideAndConcur(start, mid)
            right_half = divideAndConcur(mid + 1, end)
            return findPre(left_half, right_half)
        
        return divideAndConcur(0, len(strs) - 1)
        

File: 14-longest-common-prefix/14-longest-common-prefix.java
class Solution
{
    private static boolean canBe(String[] strs, int limit)
    {
            String pre = strs[0].substring(0, limit);
            for (int i = 1;  i < strs.length; ++i)
            {
                    if (!strs[i].startsWith(pre))
                    {
                            return false;
                    }
            }
            return true; 
    }
    
   public String longestCommonPrefix(String[] strs) 
   {
            if (strs == null || strs.length == 0) return "";
            if  (strs.length == 1) return strs[0];
            int minLen = Integer.MAX_VALUE;
            for (int i = 0; i < strs.length; i++)
            {
                    minLen = Math.min(minLen, strs[i].length());
            }
            int left = 0, right = minLen;
            while (left < right)
            {
                    int mid = left + (right - left  + 1) / 2;
                    if  (canBe(strs, mid))
                    {
                            left = mid;
                    }
                    else
                    {
                            right = mid - 1;
                    }
            }
            return strs[0].substring(0, left);   
    }
}